src/tlsh.rs

```diff
@@ -83,83 +83,42 @@ impl<
     pub fn update(&mut self, data: &[u8]) {
         // TODO: TLSH_OPTION_THREADED | TLSH_OPTION_PRIVATE
 
-        let mut j = self.data_len % RNG_SIZE;
         let mut fed_len = self.data_len;
+        let mut slide_window = u32::from_ne_bytes(self.slide_window);
```

Comment:
code is probably broken on big endian because of the `from_ne_bytes` but `to_le_bytes`
The u32 optim is clever but i'm not sure i want to rely on the RNG_SIZE being equal to 4.

I've tested a few other ways to write this:

solution 1:

```diff
@@ -84,12 +84,10 @@
         // TODO: TLSH_OPTION_THREADED | TLSH_OPTION_PRIVATE

         let mut fed_len = self.data_len;
-        let mut slide_window = u32::from_ne_bytes(self.slide_window);

         for b in data {
             let b_0 = *b;
-            let [b_1, b_2, b_3, b_4] = slide_window.to_le_bytes();
-            slide_window = (slide_window << 8) + u32::from(b_0);
+            let [b_1, b_2, b_3, b_4] = self.slide_window;

             if fed_len >= 4 {
                 for k in 0..TLSH_CHECKSUM_LEN {
@@ -115,10 +113,10 @@
                 let r = fast_b_mapping::<EFF_BUCKETS>(230, b_0, b_3, b_4);
                 self.a_bucket[usize::from(r)] += 1;
             }
+            self.slide_window = [b_0, b_1, b_2, b_3];
             fed_len += 1;
         }

-        self.slide_window = slide_window.to_ne_bytes();
         self.data_len += data.len();
     }
```

Solution 2:

```diff
@@ -84,12 +84,10 @@
         // TODO: TLSH_OPTION_THREADED | TLSH_OPTION_PRIVATE

         let mut fed_len = self.data_len;
-        let mut slide_window = u32::from_ne_bytes(self.slide_window);
+        let [mut b_1, mut b_2, mut b_3, mut b_4] = self.slide_window;

         for b in data {
             let b_0 = *b;
-            let [b_1, b_2, b_3, b_4] = slide_window.to_le_bytes();
-            slide_window = (slide_window << 8) + u32::from(b_0);

             if fed_len >= 4 {
                 for k in 0..TLSH_CHECKSUM_LEN {
@@ -115,11 +113,14 @@
                 let r = fast_b_mapping::<EFF_BUCKETS>(230, b_0, b_3, b_4);
                 self.a_bucket[usize::from(r)] += 1;
             }
+            [b_1, b_2, b_3, b_4] = [b_0, b_1, b_2, b_3];
             fed_len += 1;
         }

-        self.slide_window = slide_window.to_ne_bytes();
-        self.data_len += data.len();
+        if !data.is_empty() {
+            self.slide_window = [data[data.len() - 1], b_1, b_2, b_3];
+            self.data_len += data.len();
+        }
     }
```

When benching, i actually have the first solution very close to yours or sometimes faster, notably when enabling the fast feature.
I also have the second solution appear faster but only when *not* using the fast feature, which is peculiar.
Could you try to bench the first solution on your side and confirm that it works well? I'd prefer either one of the two to avoid the RNG_SIZE=4 limitation.

Reply:
Solution 2 is essentially where I started. The u32 intermediate hints to the compiler that the carry-over state can be stashed in a single register and the components extracted on-demand. I found a compromise with native endianness used throughout.

Reply:
I quieted my system and ran all 6 variation head-to-head.
Solution 1 was consistently faster than the others.

| Implementation| Features | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|---:|
| Solution 1 | `fast` | 21.251 ± 0.202 | 20.994 | 21.566 | 1.00 |
| Solution 2 | `fast` | 21.475 ± 0.059 | 21.399 | 21.588 | 1.01 ± 0.01 |
| `u32::to_ne_bytes` | `fast` | 23.480 ± 0.179 | 23.088 | 23.696 | 1.10 ± 0.01 |
| Solution 1 || 23.519 ± 0.315 | 23.134 | 23.995 | 1.11 ± 0.02 |
| Solution 2 | | 24.063 ± 0.656 | 23.345 | 25.239 | 1.13 ± 0.03 |
| `u32::to_ne_bytes` | | 25.825 ± 0.761 | 24.832 | 27.212 | 1.22 ± 0.04 |